src/autoschedulers/anderson2021/State.cpp

```diff
@@ -0,0 +1,1301 @@
+        return b;
+    }
+    if (a == b) {
+        return a;
+    }
+
+    // Walk the deeper one up until they're at the same depth
+    auto it_a = parent.find(a);
+    auto it_b = parent.find(b);
+    internal_assert(it_a != parent.end() && it_b != parent.end());
+    while (it_a->second.second > it_b->second.second) {
+        a = it_a->second.first;
+        it_a = parent.find(a);
+    }
+    while (it_b->second.second > it_a->second.second) {
+        b = it_b->second.first;
+        it_b = parent.find(b);
+    }
+
+    while (true) {
+        // Walk each up one
+        a = it_a->second.first;
+        b = it_b->second.first;
+        if (a == b) {
+            return a;
+        }
+        it_a = parent.find(a);
+        it_b = parent.find(b);
+        internal_assert(it_a != parent.end() && it_b != parent.end());
+    }
+
+    // unreachable
+    return nullptr;
+}
+
+bool State::has_loop_nest_without_thread_loops() const {
+    for (const auto &c : root->children) {
+        if (c->gpu_label != block) {
+            continue;
+        }
+
+        for (const auto &block_c : c->children) {
+            if (!block_c->all_paths_to_leaves_have_thread_loop()) {
+                return true;
+            }
+        }
+    }
+
+    return false;
+}
+
+bool State::has_compute_root_loops_without_blocks() const {
+    for (const auto &c : root->children) {
+        if (c->gpu_label == none) {
+            return true;
+        }
+    }
+
+    return false;
+}
+
+void State::FeatureLoopNestMutator::operator()(LoopNest *new_loop_nest) const {
+    split_compute_root_loops(new_loop_nest);
+    add_outer_thread_loops(new_loop_nest);
+}
+
+// In phase 2, any compute_root loop marked 'none' will be split into
+// blocks, threads, and serial loops. To enable the cost model to make a
+// meaningful prediction on these pre-split loops, we assume a split into
+// blocks and threads with a single full warp (if possible)
+void State::FeatureLoopNestMutator::split_compute_root_loops(LoopNest *loop_nest) const {
+    if (!loop_nest || !loop_nest->is_root()) {
+        return;
+    }
+
+    for (auto it = loop_nest->children.rbegin(); it != loop_nest->children.rend(); ++it) {
+        auto &c = *it;
+        if (c->gpu_label != none) {
+            continue;
+        }
+
+        int vectorized_loop_index = c->vectorized_loop_index;
+
+        if (c->size.empty()) {
+            continue;
+        }
+
+        // Make the vectorized dimension of the inner loop 32 (or as
+        // close as possible)
+        int64_t inner_extent = std::min(c->size[vectorized_loop_index], (int64_t)32);
+
+        if (c->stage->index == 0) {
+            vector<int64_t> tiling(c->node->dimensions, 1);
+
+            // Split into parallelized and serial
+            c = c->parallelize_in_tiles(tiling, loop_nest, params, target, true, false);
+
+            if (vectorized_loop_index >= 0) {
+                tiling[vectorized_loop_index] = inner_extent;
+            }
```

Comment:
It seems that it is not checked for whether `vectorized_loop_index == -1` here in line 222 (and it seems that `-1` is a legal value here because of the subsequent check in line 230), and leads to a crash when running `test/autoschedulers/anderson2021/test.cpp`. I changed
```C++
int64_t inner_extent = std::min(c->size[vectorized_loop_index], (int64_t)32);
```
to
```C++
int64_t inner_extent = vectorized_loop_index != -1 ? std::min(c->size[vectorized_loop_index], (int64_t)32) : (int64_t)32;
```
to fix this but I am not sure if this is right.